proposals/0000-row-polymorphism.rst

```diff
@@ -48,86 +48,184 @@ There are mainly two reasons why I open this proposal now:
-  printName :: { name :: String | r } -> IO ()
-  printName { name } = putStrLn name
-  -- equivalent to
-  printName2 :: RowCons "name" String r r0 => Record r0 -> IO ()
-  printName2 r = let name = get #name r
-                 in putStrLn name
 
-  -- defined in Data.Record
-  (@.) :: RowCons s ty _ r => Record r -> Proxy s -> ty
+  --| Asserts that the row lacks the label
+  class RowLacks label row
+
+  --| Removes duplicate labels from the row
+  class RowNub original nubbed | original -> nubbed
 
-  -- pseudo typeclass, in reality builtin constraint
-  class RowCons (s :: Symbol) (ty :: k) (r1 :: Row k) (r2 :: Row k) | s ty r1 -> r2, s r2 -> ty r1
+  --| Merges to rows (left-biased), includes duplicated labels
+  class RowUnion left right union
+        | left right -> union, right union -> left, left union -> right
 
-This means, that ``(foo :: Int, bar :: Y)`` has kind ``Row Type`` (or ``Row *``), ``Record (foo :: Int, bar Y)`` has kind ``Type`` (or ``*``) and ``{ foo :: Int, bar :: Y }`` is syntactic sugar for ``Record (foo :: Int, bar :: Y)`` so it also has kind ``Type`` (or ``*``).
+2. Records
 
-The desugaring of types in the form ``{ foo :: Foo, bar :: Bar | r}`` would work like this - note that those types are desugared after type family expansion
+As a second step, a new type for records is introduced:
+
+2.1 Syntax
 
 .. code-block:: haskell
 
-  type family WithId (r :: Row Type) :: Type where
-      WithId r = { id :: Int | r }
+  data Record (r :: Row Type) = Record (SmallArray# Any)
 
-  f :: { | r } -> WithId r
-  -- desugaring steps
-  f :: { | r } -> { id :: Int | r }
-  f :: forall r. { | r } -> { id :: Int | r }
-  f :: forall r. Record ( | r ) -> Record (id :: Int | r)
-  f :: forall r. Record r -> Record (id :: Int | r)
-  f :: forall r r0. RowCons "id" Int r r0 => Record r -> Record ( | r0)
-  f :: forall r r0. RowCons "id" Int r r0 => Record r -> Record r0
+The type level syntax of a record is very similar to rows:
+
+.. code-block:: haskell
 
-The newly introduced type variables and constraints will always be added to the ``forall`` that quantifies the variable of the open row, so for a higher kinded and a RankN type:
+  recordType ::= '{' [rowFields] ['|' (typeVar | row)] '}'
+
+This is just syntactic sugar for the record type and its row:
 
 .. code-block:: haskell
 
-  f :: Maybe { name :: String | r } -> String
-  -- desugaring steps
-  f :: forall r. Maybe { name :: String | r } -> String
-  f :: forall r. Maybe (Record (name :: String | r)) -> String
-  f :: forall r r0. RowCons "name" String r r0 => Maybe (Record ( | r0)) -> String
-  f :: forall r r0. RowCons "name" String r r0 => Maybe (Record r0) -> String
+  { foo :: Foo, bar :: Bar | r } ~ Record (foo :: Foo, bar :: Bar | r)
+
+The value level syntax is similar to the current syntax for records and the syntax in other languages:
+
+.. code-block:: haskell
+
+  record ::= '{' [recordFields] '}'
+  recordFields ::= label ':' expression [',' recordFields]
```

Comment:
`:` already has a value-level meaning, it would probably be better to retain the existing record syntax wholesale with

```
recordFields ::= label '=' expression [',' recordFields]
```

Reply:
Actually, I withdraw the suggestion to retain the existing record syntax, as you might need to use existing records and polymorphic records in the same module. Still, I think reusing `:` is liable to cause confusion, even though it doesn't seem to be syntactically ambiguous.

Reply:
I agree

Reply:
I agree with `=`. You won't be able to mix the old and the new syntax in the same file, enabling the extension will fill disable the old syntax